hail/src/main/scala/is/hail/io/tabix/TabixReader.scala

```diff
@@ -0,0 +1,346 @@
+package is.hail.io.tabix
+
+import is.hail.HailContext
+import is.hail.io.compress.BGzipInputStream
+import is.hail.utils._
+
+import htsjdk.tribble.util.{ParsingUtils, TabixUtils}
+import org.apache.{hadoop => hd}
+import org.apache.hadoop.io.compress.SplitCompressionInputStream
+
+import java.io.InputStream
+import java.nio.{ByteBuffer, ByteOrder}
+import java.util.Arrays
+
+import scala.collection.mutable.HashMap
+import scala.language.implicitConversions
+
+// Helper data classes
+
+case class Tabix(
+  val format: Int,
+  val colSeq: Int,
+  val colBeg: Int,
+  val meta: Int,
+  val seqs: Array[String],
+  val chr2tid: HashMap[String, Int],
+  val indicies: Array[(HashMap[Int, Array[TbiPair]], Array[Long])]
+)
+
+case class TbiPair(var _1: Long, var _2: Long) extends java.lang.Comparable[TbiPair] {
+  @Override
+  def compareTo(other: TbiPair) = TbiOrd.compare(this, other)
+}
+
+object TbiPair {
+  implicit def tup2Pair(t: (Long, Long)): TbiPair = new TbiPair(t._1, t._2)
+}
+
+object TbiOrd extends Ordering[TbiPair] {
+  def compare(u: TbiPair, v: TbiPair) = if (u._1 == v._1) {
+    0
+  } else if ((u._1 < v._1) ^ (u._1 < 0) ^ (v._1 < 0)) {
+    -1
+  } else {
+    1
+  }
+
+  def less64(u: Long, v: Long) = (u < v) ^ (u < 0) ^ (v < 0)
+}
+
+// Tabix reader
+// Tabix file format is described here: https://samtools.github.io/hts-specs/tabix.pdf
+
+object TabixReader {
+  val MaxBin: Int = 37450
+  val TadLidxShift: Int = 14
+  val DefaultBufferSize: Int = 1000
+  val Magic: Array[Byte] = Array(84, 66, 73, 1) // "TBI\1"
+
+  def readInt(is: InputStream): Int = {
+    val buf = new Array[Byte](4)
+    is.read(buf)
+    ByteBuffer.wrap(buf).order(ByteOrder.LITTLE_ENDIAN).getInt
+  }
+
+  def readLong(is: InputStream): Long = {
+    val buf = new Array[Byte](8)
+    is.read(buf)
+    ByteBuffer.wrap(buf).order(ByteOrder.LITTLE_ENDIAN).getLong
+  }
+
+  def readLine(is: InputStream): String = readLine(is, DefaultBufferSize)
+
+  def readLine(is: InputStream, initialBufferSize: Int): String = {
+    val buf = new StringBuffer(initialBufferSize)
+    var c = is.read()
+    while (c >= 0 && c != '\n') {
+      buf.append(c.asInstanceOf[Char])
+      c = is.read()
+    }
+    buf.toString()
+  }
+
+}
+
+class TabixReader(val filePath: String, private val idxFilePath: Option[String]) {
+  import TabixReader._
+
+  val indexPath: String = idxFilePath match {
+    case None => ParsingUtils.appendToPath(filePath, TabixUtils.STANDARD_INDEX_EXTENSION)
+    case Some(s) => {
+      if (s.endsWith(".tbi"))
+        s
+      else
+        fatal(s"unknown file extension for tabix index: ${s}")
+    }
+  }
+
+  def this(filePath: String) = this(filePath, None)
```

Comment:
style: prefer default arguments in class constructor above to constructor overloading.